File: src/checkerTypes.ts
// The type model used by the checker. Java types: primitives, class/interface
// types (with type arguments), arrays, type variables, wildcards, intersections,
// the null type and an error type for the unknown/unresolved (so analysis
// degrades gracefully instead of throwing or reporting false errors).

import type { Symbol } from "./types.ts";

export const enum TypeKind {
  Primitive,
  Class,
  Array,
  TypeVariable,
  Wildcard,
  Intersection,
  Null,
  Error,
}

export interface PrimitiveType {
  readonly kind: TypeKind.Primitive;
  readonly name: string; // int, long, boolean, ..., void
}

export interface ClassType {
  readonly kind: TypeKind.Class;
  readonly symbol: Symbol;
  readonly typeArguments: readonly Type[];
}

export interface ArrayType {
  readonly kind: TypeKind.Array;
  readonly elementType: Type;
}

export interface TypeVariable {
  readonly kind: TypeKind.TypeVariable;
  readonly symbol: Symbol;
}

export interface WildcardType {
  readonly kind: TypeKind.Wildcard;
  readonly bound?: Type;
  readonly isExtends: boolean;
  readonly isSuper: boolean;
}

export interface IntersectionType {
  readonly kind: TypeKind.Intersection;
  readonly types: readonly Type[];
}

export interface NullType {
  readonly kind: TypeKind.Null;
}

export interface ErrorType {
  readonly kind: TypeKind.Error;
}

export type Type =
  | PrimitiveType
  | ClassType
  | ArrayType
  | TypeVariable
  | WildcardType
  | IntersectionType
  | NullType
  | ErrorType;

export const errorType: ErrorType = { kind: TypeKind.Error };
export const nullType: NullType = { kind: TypeKind.Null };

const primitiveCache = new Map<string, PrimitiveType>();
export function primitiveType(name: string): PrimitiveType {
  let t = primitiveCache.get(name);
  if (!t) {
    t = { kind: TypeKind.Primitive, name };
    primitiveCache.set(name, t);
  }
  return t;
}

export function classType(symbol: Symbol, typeArguments: readonly Type[] = []): ClassType {
  return { kind: TypeKind.Class, symbol, typeArguments };
}

export function arrayType(elementType: Type): ArrayType {
  return { kind: TypeKind.Array, elementType };
}

export function typeVariable(symbol: Symbol): TypeVariable {
  return { kind: TypeKind.TypeVariable, symbol };
}

export function isError(type: Type): boolean {
  return type.kind === TypeKind.Error;
}

/** Human-readable form for hover/diagnostics. */
export function typeToString(type: Type): string {
  switch (type.kind) {
    case TypeKind.Primitive:
      return type.name;
    case TypeKind.Class: {
      const name = type.symbol.escapedName;
      if (type.typeArguments.length === 0) return name;
      return `${name}<${type.typeArguments.map(typeToString).join(", ")}>`;
    }
    case TypeKind.Array:
      return `${typeToString(type.elementType)}[]`;
    case TypeKind.TypeVariable:
      return type.symbol.escapedName;
    case TypeKind.Wildcard:
      if (type.isExtends && type.bound) return `? extends ${typeToString(type.bound)}`;
      if (type.isSuper && type.bound) return `? super ${typeToString(type.bound)}`;
      return "?";
    case TypeKind.Intersection:
      return type.types.map(typeToString).join(" & ");
    case TypeKind.Null:
      return "null";
    default:
      return "<error>";
  }
}

File: togo/internal/compiler/checker_types.go
package compiler

// The type model used by the checker. Java types: primitives, class/interface
// types (with type arguments), arrays, type variables, wildcards, intersections,
// the null type and an error type for the unknown/unresolved (so analysis
// degrades gracefully instead of throwing or reporting false errors).
// Port of src/compiler/checkerTypes.ts.

import "strings"

type TypeKind int

const (
	TypeKindPrimitive TypeKind = iota
	TypeKindClass
	TypeKindArray
	TypeKindTypeVariable
	TypeKindWildcard
	TypeKindIntersection
	TypeKindNull
	TypeKindError
)

// Type is a Java type in the checker's model. The active fields depend on Kind.
type Type struct {
	Kind TypeKind

	Name          string  // Primitive: int, long, boolean, ..., void
	Symbol        *Symbol // Class, TypeVariable
	TypeArguments []*Type // Class
	ElementType   *Type   // Array
	// Bound is the leftmost declared bound of a TypeVariable (filled in lazily by
	// the checker, since it may reference the variable itself), or a Wildcard's bound.
	Bound     *Type
	IsExtends bool    // Wildcard
	IsSuper   bool    // Wildcard
	Types     []*Type // Intersection

	// Nullness is the jspecify nullness facet (nikeee/cappu#25), attached by
	// resolveType only when nullness checking is enabled and read only by the
	// nullness checks - typeToString, assignability and the emitter ignore it.
	Nullness Nullness
}

// nullnessOf returns the jspecify nullness facet of a type ("" when unknown).
func nullnessOf(t *Type) Nullness {
	return t.Nullness
}

// withNullness returns a copy of t with its nullness facet set (nikeee/cappu#25).
// Only reference types carry nullness; a no-op for primitives, null, error and "".
func withNullness(t *Type, nullness Nullness) *Type {
	if nullness == "" {
		return t
	}
	switch t.Kind {
	case TypeKindClass, TypeKindArray, TypeKindTypeVariable:
		clone := *t
		clone.Nullness = nullness
		return &clone
	}
	return t
}

var (
	errorType = &Type{Kind: TypeKindError}
	nullType  = &Type{Kind: TypeKindNull}
)

var primitiveCache = map[string]*Type{}

func primitiveType(name string) *Type {
	if t, ok := primitiveCache[name]; ok {
		return t
	}
	t := &Type{Kind: TypeKindPrimitive, Name: name}
	primitiveCache[name] = t
	return t
}

func classType(symbol *Symbol, typeArguments []*Type) *Type {
	return &Type{Kind: TypeKindClass, Symbol: symbol, TypeArguments: typeArguments}
}

func arrayType(elementType *Type) *Type {
	return &Type{Kind: TypeKindArray, ElementType: elementType}
}

func typeVariable(symbol *Symbol) *Type {
	return &Type{Kind: TypeKindTypeVariable, Symbol: symbol}
}

func isError(t *Type) bool {
	return t.Kind == TypeKindError
}

// TypeToString renders a human-readable form for hover/diagnostics (exported
// for the language-services layer).
func TypeToString(t *Type) string { return typeToString(t) }

// typeToString renders a human-readable form for hover/diagnostics.
func typeToString(t *Type) string {
	switch t.Kind {
	case TypeKindPrimitive:
		return t.Name
	case TypeKindClass:
		name := t.Symbol.EscapedName
		if len(t.TypeArguments) == 0 {
			return name
		}
		parts := make([]string, len(t.TypeArguments))
		for i, a := range t.TypeArguments {
			parts[i] = typeToString(a)
		}
		return name + "<" + strings.Join(parts, ", ") + ">"
	case TypeKindArray:
		return typeToString(t.ElementType) + "[]"
	case TypeKindTypeVariable:
		return t.Symbol.EscapedName
	case TypeKindWildcard:
		if t.IsExtends && t.Bound != nil {
			return "? extends " + typeToString(t.Bound)
		}
		if t.IsSuper && t.Bound != nil {
			return "? super " + typeToString(t.Bound)
		}
		return "?"
	case TypeKindIntersection:
		parts := make([]string, len(t.Types))
		for i, x := range t.Types {
			parts[i] = typeToString(x)
		}
		return strings.Join(parts, " & ")
	case TypeKindNull:
		return "null"
	default:
		return "<error>"
	}
}
